server/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -682,21 +683,21 @@ public int size() {
 
     /** Returns the fully qualified setting names contained in this settings object. */
     public Set<String> keySet() {
-        if (keys.get() == null) {
-            synchronized (keys) {
-                // Check that the keys are still null now that we have acquired the lock
-                if (keys.get() == null) {
-                    if (secureSettings == null) {
-                        keys.set(settings.keySet());
-                    } else {
-                        Stream<String> stream = Stream.concat(settings.keySet().stream(), secureSettings.getSettingNames().stream());
-                        // uniquify, since for legacy reasons the same setting name may exist in both
-                        keys.set(Collections.unmodifiableSet(stream.collect(Collectors.toSet())));
-                    }
-                }
-            }
+        final Set<String> keySet = keys;
+        if (keySet != null) {
+            return keySet;
+        }
+        final Set<String> newKeySet;
+        if (secureSettings == null) {
+            newKeySet = settings.keySet();
+        } else {
+            // uniquify, since for legacy reasons the same setting name may exist in both
+            final Set<String> merged = new HashSet<>(settings.keySet());
```

Comment:
I'm guessing creating a merged Set using addAll is faster than doing stream concat, which is why we switched the approach? If that's the case, can similar approach be done on merging the names in names() above? 

I wonder though, what if the keySet and the secure names were sorted sets, I think we might break ordering by selecting HashSet if the consumer expected it.

Reply:
I think the previous code using `java.util.stream.Collectors#toSet` guarantees no ordering either (it's also based on `HashSet` as far as I can see) so I don't think we're changing anything here on that front.